lib/CGameState.cpp

```diff
@@ -3276,53 +3273,162 @@ TeamState::TeamState()
 
 void CPathfinder::initializeGraph()
 {
+	int3 pos;
 	CGPathNode ***graph = out.nodes;
-	for(size_t i=0; i < out.sizes.x; ++i)
+	for(pos.x=0; pos.x < out.sizes.x; ++pos.x)
 	{
-		for(size_t j=0; j < out.sizes.y; ++j)
+		for(pos.y=0; pos.y < out.sizes.y; ++pos.y)
 		{
-			for(size_t k=0; k < out.sizes.z; ++k)
+			for(pos.z=0; pos.z < out.sizes.z; ++pos.z)
 			{
-				curPos = int3(i,j,k);
-				const TerrainTile *tinfo = &gs->map->getTile(int3(i, j, k));
-				CGPathNode &node = graph[i][j][k];
-
-				node.accessible = evaluateAccessibility(tinfo);
+				const TerrainTile *tinfo = &gs->map->getTile(pos);
+				CGPathNode &node = graph[pos.x][pos.y][pos.z];
+				node.accessible = evaluateAccessibility(pos, tinfo);
 				node.turns = 0xff;
 				node.moveRemains = 0;
-				node.coord.x = i;
-				node.coord.y = j;
-				node.coord.z = k;
-                node.land = tinfo->terType != ETerrainType::WATER;
+				node.coord = pos;
+				node.land = tinfo->terType != ETerrainType::WATER;
 				node.theNodeBefore = nullptr;
 			}
 		}
 	}
 }
 
+void CPathfinder::addNeighbours(const int3 &coord)
+{
+	neighbours.clear();
+	ct = &gs->map->getTile(coord);
+
+	std::vector<int3> tiles;
+	gs->getNeighbours(*ct, coord, tiles, boost::logic::indeterminate, !cp->land);
+	sTileObj = ct->topVisitableObj(coord == CGHeroInstance::convertPosition(hero->pos, false));
+	if(sTileObj)
+	{
+		for(int3 tile: tiles)
+		{
+			if(canMoveBetween(tile, sTileObj->visitablePos()))
+				neighbours.push_back(tile);
+		}
+	}
+	else
+		vstd::concatenate(neighbours, tiles);
+}
+
+void CPathfinder::addTeleportExits(bool noTeleportExcludes)
+{
+	assert(sTileObj);
+
+	neighbours.clear();
+	auto isAllowedTeleportEntrance = [&](const CGTeleport * obj) -> bool
+	{
+		if(!gs->isTeleportEntrancePassable(obj, hero->tempOwner))
+			return false;
+
+		if(noTeleportExcludes)
+			return true;
+
+		auto whirlpool = dynamic_cast<const CGWhirlpool *>(obj);
+		if(whirlpool)
+		{
+			if(addTeleportWhirlpool(whirlpool))
+				return true;
+		}
+		else if(addTeleportTwoWay(obj) || addTeleportOneWay(obj) || addTeleportOneWayRandom(obj))
+			return true;
+
+		return false;
+	};
+
+	const CGTeleport *sTileTeleport = dynamic_cast<const CGTeleport *>(sTileObj);
+	if(isAllowedTeleportEntrance(sTileTeleport))
+	{
+		for(auto objId : gs->getTeleportChannelExits(sTileTeleport->channel, hero->tempOwner))
+		{
+			auto obj = getObj(objId);
+			if(CGTeleport::isExitPassable(gs, hero, obj))
+				neighbours.push_back(obj->visitablePos());
+		}
+	}
+}
+
+int3 CPathfinder::getSourceGuardPosition()
+{
+	return gs->map->guardingCreaturePositions[cp->coord.x][cp->coord.y][cp->coord.z];
+}
+
+bool CPathfinder::isSourceGuarded()
+{
+	//TODO: find out why exactly source can't be guarded if hero on it. Why this quirk was nessesaary in first place? Map where hero start on guarded tile?
```

Comment:
Hero starts on guarded tile, hero teleports on guarded tile (via teleporter), monsters spawned near hero on new month. This is normal situation and not some exception from the rules.